src/lib/support/jsontlv/TlvToJson.cpp

```diff
@@ -0,0 +1,371 @@
+ * Given a TLVReader positioned at TLV structure this function:
+ *   - enters structure
+ *   - converts all elements of a structure into JSON object representation
+ *   - exits structure
+ */
+CHIP_ERROR TlvStructToJson(TLV::TLVReader & reader, Json::Value & jsonObj)
+{
+    CHIP_ERROR err;
+    TLV::TLVType containerType;
+
+    ReturnErrorOnFailure(reader.EnterContainer(containerType));
+
+    while ((err = reader.Next()) == CHIP_NO_ERROR)
+    {
+        TLV::Tag tag = reader.GetTag();
+        VerifyOrReturnError(TLV::IsContextTag(tag) ||
+                                (TLV::IsProfileTag(tag) && (TLV::ProfileIdFromTag(tag) == TLV::kCommonProfileId)),
+                            CHIP_ERROR_INVALID_TLV_TAG);
+        KeyContext context(static_cast<chip::FieldId>(TLV::TagNumFromTag(tag)));
+
+        // Recursively convert to JSON the encompassing item within the struct.
+        ReturnErrorOnFailure(TlvToJson(reader, context, jsonObj));
+    }
+
+    VerifyOrReturnError(err == CHIP_END_OF_TLV, err);
+    return reader.ExitContainer(containerType);
+}
+
+CHIP_ERROR TlvToJson(TLV::TLVReader & reader, KeyContext context, Json::Value & jsonObj)
+{
+    switch (reader.GetType())
+    {
+    case TLV::kTLVType_UnsignedInteger: {
+        uint64_t v;
+        ReturnErrorOnFailure(reader.Get(v));
+        context.SetElementType(kElementTypeUInt);
+        if (CanCastTo<uint32_t>(v))
+        {
+            InsertKeyValue(jsonObj, context, v);
+        }
+        else
+        {
+            InsertKeyValue(jsonObj, context, std::to_string(v));
+        }
+        break;
+    }
+
+    case TLV::kTLVType_SignedInteger: {
+        int64_t v;
+        ReturnErrorOnFailure(reader.Get(v));
+        context.SetElementType(kElementTypeInt);
+        if (CanCastTo<int32_t>(v))
+        {
+            InsertKeyValue(jsonObj, context, v);
+        }
+        else
+        {
+            InsertKeyValue(jsonObj, context, std::to_string(v));
+        }
+        break;
+    }
+
+    case TLV::kTLVType_Boolean: {
+        bool v;
+        ReturnErrorOnFailure(reader.Get(v));
+        context.SetElementType(kElementTypeBool);
+        InsertKeyValue(jsonObj, context, v);
+        break;
+    }
+
+    case TLV::kTLVType_FloatingPointNumber: {
+        double v;
+        ReturnErrorOnFailure(reader.Get(v));
+        VerifyOrReturnError(v != std::numeric_limits<double>::infinity() && v != -std::numeric_limits<double>::infinity(),
+                            CHIP_ERROR_INVALID_TLV_ELEMENT);
+        if (reader.IsElementDouble())
+        {
+            context.SetElementType(kElementTypeDouble);
+        }
+        else
+        {
+            context.SetElementType(kElementTypeFloat);
+        }
+        InsertKeyValue(jsonObj, context, v);
+        break;
+    }
+
+    case TLV::kTLVType_ByteString: {
+        ByteSpan span;
+        ReturnErrorOnFailure(reader.Get(span));
+        VerifyOrReturnError(span.size() < kMaxStringLen, CHIP_ERROR_INVALID_TLV_ELEMENT);
```

Comment:
Why do we actually care about kMaxStringLen?  I don't see us depending on it anywhere...

Reply:
agree. this was carried over from the older implementation. I removed that restriction.